### PR TITLE
Simple match%hw support

### PR DESCRIPTION
**Pull request by andrewray**

### CODE REVIEW

File: src/ppx_hardcaml.ml [hunk 1]
```diff
@@ -104,60 +93,166 @@ let const_mapper ~signed = function
...
+      `Recurse [%expr bit [%e s] [%e i]]
     (* if/then/else construct *)
     | [%expr [%hw if [%e? cnd] then [%e? e0] else [%e? e1]]] ->
-      Some [%expr mux2 [%e cnd] [%e e0 ++ e1] [%e e1 ++ e0]]
+      `Recurse (app [%expr mux2 [%e cnd]] e0 e1)
+    (* mux *)
+    | [%expr [%hw [%e? { pexp_desc=Pexp_match(sel, cases); pexp_loc=loc }]]] ->
+      let sel = m.expr m sel in
+      let cases = List.map (fun c -> { c with pc_rhs = m.expr m c.pc_rhs }) cases in
+      `Return (match_mapper ~loc resize sel cases)
```
Comment (xguerin) [inline src/ppx_hardcaml.ml]: If I am not mistaken, this algorithm does not recurse over case expression. I don't think that is what we want, right ?

File: src/ppx_hardcaml.ml [hunk 2]
```diff
@@ -104,60 +93,166 @@ let const_mapper ~signed = function
...
+  in
+  check_unique_cases cases;
+
+  (* turn cases into an [(int*signal) list] expression *)
+  let cases = List.fold_right 
+    ~f:(fun (x,y) l -> [%expr ([%e int x], [%e y]) :: [%e l]]) cases ~init:[%expr []] 
+  in
+
+  [%expr
+    (fun resize sel cases default -> (* this should be in a library *)
```
Comment (andrewray) [inline src/ppx_hardcaml.ml]: yes, it could.

**Comment by andrewray** [comment]
On Mon, Dec 19, 2016 at 6:38 PM, Xavier Guérin <notifications@github.com>
wrote:

> *@xguerin* commented on this pull request.
> ------------------------------
>
> In src/ppx_hardcaml.ml
> <https://github.com/xguerin/ppx_hardcaml/pull/16#pullrequestreview-13613759>
> :
>
> >      (* if/then/else construct *)
>      | [%expr [%hw if [%e? cnd] then [%e? e0] else [%e? e1]]] ->
> -      Some [%expr mux2 [%e cnd] [%e e0 ++ e1] [%e e1 ++ e0]]
> +      `Recurse (app [%expr mux2 [%e cnd]] e0 e1)
> +    (* mux *)
> +    | [%expr [%hw [%e? { pexp_desc=Pexp_match(sel, cases); pexp_loc=loc }]]] ->
> +      let sel = m.expr m sel in
> +      let cases = List.map (fun c -> { c with pc_rhs = m.expr m c.pc_rhs }) cases in
> +      `Return (match_mapper ~loc resize sel cases)
>
> If I am not mistaken, this algorithm does not recurse over case
> expression. I don't think that is what we want, right ?
>


It recurses explicitly on the select and the rhs part of each case, so you
can write (a,b,c,d are signals)

```
let%hw f a b c d =
  match%hw a+b with
  | 10 -> c+d
  | _ -> c-d
```

It's done within the branch here rather than recursing on the whole (new)
expression because otherwise it recurses into that big generated function
and breaks things.

**Comment by xguerin** [comment]
Right, I realized that just after posting the comment and then deleted it. I did not realize you would be able to see it ;)

**Comment by xguerin** [comment]
Gosh, your last update completely eluded me. My apologies :/ I believe you have a new version of HardCaml with the new `matches` and `resize_op2` functions. Could you please update the `opam` file accordingly ?

**Comment by andrewray** [comment]
I'm planning on releasing new versions of hardcaml and friends this week (possibly tomorrow).  I'll let you know when it's out so we can test this change and get it merged.

**Comment by andrewray** [comment]
hardcaml.1.2.0 just got merged, so I should be able to test this properly tomorrow.

**Comment by xguerin** [comment]
Awesome. Let me know when you plan to release it in OPAM. I'll create a release for this package as well.

**Comment by andrewray** [comment]
The required support is in the latest hardcaml version in OPAM (v1.2.0)

**Comment by xguerin** [comment]
Excellent.

**Pull request merged**